File: 01-programming-languages/JavaScript/01-fundamentals/03-control-flow/loops.js
/*
 * JavaScript Control Flow: Loops
 * for, while, for...of, for...in
 */

// ============================================
// for Loop
// ============================================

console.log("for loop:");
for (let i = 0; i < 5; i++) {
    console.log(`Iteration ${i}`);
}

// ============================================
// while Loop
// ============================================

console.log("\nwhile loop:");
let count = 0;
while (count < 5) {
    console.log(`Count: ${count}`);
    count++;
}

// ============================================
// do...while Loop
// ============================================

console.log("\ndo...while loop:");
let num = 0;
do {
    console.log(`Number: ${num}`);
    num++;
} while (num < 5);

// ============================================
// for...of Loop (iterates over values)
// ============================================

console.log("\nfor...of loop:");
const fruits = ["apple", "banana", "cherry"];

for (const fruit of fruits) {
    console.log(`Fruit: ${fruit}`);
}

// Works with strings
const word = "Hello";
for (const char of word) {
    console.log(`Character: ${char}`);
}

// Works with Maps
const map = new Map([["a", 1], ["b", 2]]);
for (const [key, value] of map) {
    console.log(`${key}: ${value}`);
}

// ============================================
// for...in Loop (iterates over keys/indices)
// ============================================

console.log("\nfor...in loop:");
const person = { name: "Alice", age: 30, city: "NYC" };

for (const key in person) {
    console.log(`${key}: ${person[key]}`);
}

// with arrays (indices)
const colors = ["red", "green", "blue"];
for (const index in colors) {
    console.log(`Index ${index}: ${colors[index]}`);
}

// ============================================
// break and continue
// ============================================

console.log("\nbreak example:");
for (let i = 0; i < 10; i++) {
    if (i === 5) break; // Exit loop when i is 5
    console.log(i);
}

console.log("\ncontinue example:");
for (let i = 0; i < 10; i++) {
    if (i % 2 === 0) continue; // Skip even numbers
    console.log(i);
}

// ============================================
// Labeled Statements
// ============================================

console.log("\nlabeled loop:");
outer: for (let i = 0; i < 3; i++) {
    for (let j = 0; j < 3; j++) {
        if (i === 1 && j === 1) break outer;
        console.log(`${i}, ${j}`);
    }
}

// ============================================
// Array Iteration Methods (Functional Approach)
// ============================================

const numbers = [1, 2, 3, 4, 5];

// forEach - no break/continue possible
numbers.forEach((num, index) => {
    console.log(`Index ${index}: ${num}`);
});

// map - creates new array
const doubled = numbers.map(num => num * 2);
console.log("Doubled:", doubled);

// filter - creates new array with matching elements
const evens = numbers.filter(num => num % 2 === 0);
console.log("Evens:", evens);

File: 01-programming-languages/C#/01-fundamentals/03-control-flow/loops.cs
// C# Control Flow - Loops

using System;

namespace LoopsDemo
{
    class Program
    {
        static void Main(string[] args)
        {
            // for loop
            Console.WriteLine("For loop:");
            for (int i = 0; i < 5; i++)
            {
                Console.Write($"{i} ");
            }
            Console.WriteLine();
            
            // Nested for loop
            Console.WriteLine("Multiplication table:");
            for (int i = 1; i <= 3; i++)
            {
                for (int j = 1; j <= 3; j++)
                {
                    Console.Write($"{i * j,4}");
                }
                Console.WriteLine();
            }
            
            // while loop
            Console.WriteLine("While loop:");
            int count = 0;
            while (count < 5)
            {
                Console.Write($"{count} ");
                count++;
            }
            Console.WriteLine();
            
            // do-while loop
            Console.WriteLine("Do-while loop:");
            int num = 0;
            do
            {
                Console.Write($"{num} ");
                num++;
            } while (num < 5);
            Console.WriteLine();
            
            // foreach loop
            Console.WriteLine("Foreach loop:");
            string[] fruits = { "Apple", "Banana", "Cherry" };
            foreach (string fruit in fruits)
            {
                Console.Write($"{fruit} ");
            }
            Console.WriteLine();
            
            // foreach with index
            Console.WriteLine("Foreach with index:");
            for (int i = 0; i < fruits.Length; i++)
            {
                Console.WriteLine($"{i}: {fruits[i]}");
            }
            
            // break statement
            Console.WriteLine("Break example:");
            for (int i = 0; i < 10; i++)
            {
                if (i == 5)
                {
                    break;
                }
                Console.Write($"{i} ");
            }
            Console.WriteLine();
            
            // continue statement
            Console.WriteLine("Continue example:");
            for (int i = 0; i < 10; i++)
            {
                if (i % 2 == 0)
                {
                    continue;
                }
                Console.Write($"{i} ");
            }
            Console.WriteLine();
            
            // goto statement (use sparingly)
            Console.WriteLine("Goto example:");
            for (int i = 0; i < 10; i++)
            {
                if (i == 5)
                {
                    goto Found;
                }
            }
            Console.WriteLine("Not found");
            goto End;
            
            Found:
            Console.WriteLine("Found 5");
            
            End:
            Console.WriteLine("End of program");
            
            // Infinite loop with break
            Console.WriteLine("Infinite loop with break:");
            int counter = 0;
            while (true)
            {
                if (counter >= 5)
                {
                    break;
                }
                Console.Write($"{counter} ");
                counter++;
            }
            Console.WriteLine();
            
            Console.WriteLine("Loops example running");
        }
    }
}